### PR TITLE
Add GrooVAE.

### PR DESCRIPTION
**Pull request by adarob**
@tambien @jrgillick

### CODE REVIEW

File: music/demos/build-demos.sh [hunk 1]
```diff
@@ -23,7 +23,7 @@
 if [[ -n $1 ]]; then
   demos=( "${1}" )
 else
-  demos=( "music_rnn.ts" "music_vae.ts" "multitrack.ts" "player.ts" "callback_player.ts" "visualizer.ts" "transcription.ts" "recorder.ts" "piano_genie.ts" )
+  demos=( "music_rnn.ts" "music_vae.ts" "multitrack.ts" "player.ts" "callback_player.ts" "visualizer.ts" "transcription.ts" "recorder.ts" "piano_genie.ts" "groovae.ts" )
```
Comment (notwaldorf) [inline music/demos/build-demos.sh]: can we like change this to be *.ts or something? it's getting silly.

File: music/demos/index.html [hunk 2]
```diff
@@ -40,6 +40,7 @@ <h3>
         <li><a href="music_rnn.html">Continue a NoteSequence using MusicRNN</a></li>
         <li><a href="music_vae.html">Generate new NoteSequences using MusicVAE</a></li>
         <li><a href="multitrack.html">Multitrack MusicVAE</a></li>
+        <li><a href="groovae.html">MusicVAE for drum performances</a></li>
```
Comment (notwaldorf) [inline music/demos/index.html]: add something about humanizing here

File: music/demos/build-demos.sh (hunk 1 above)
Comment (adarob) [reply music/demos/build-demos.sh]: Done.

File: music/src/core/data.ts [hunk 3]
```diff
@@ -136,8 +143,8 @@ export abstract class DataConverter {
   readonly numSegments: number;              // Number of steps for conductor.
   abstract readonly depth: number;           // Size of final output dimension.
   abstract readonly endTensor: tf.Tensor1D;  // Tensor marking segment end.
-  abstract readonly NUM_SPLITS: number;  // Const number of conductor splits.
-  abstract readonly SEGMENTED_BY_TRACK: boolean;  // Segments are tracks.
+  readonly NUM_SPLITS: number = 0;  // Const number of conductor splits.
```
Comment (notwaldorf) [inline music/src/core/data.ts]: if these aren't abstract, should they be moved below with their brethren NOTE_OFF, etc?

File: music/src/core/data.ts [hunk 4]
```diff
@@ -799,3 +800,170 @@ export class MultitrackConverter extends DataConverter {
...
+ * offset value, which refers to the relative distance from the nearest
+ * quantized step.
+ *
+ * Hits are binary [0, 1].
+ * Velocities are continuous values in [0, 1].
+ * Offsets are continuous values in [-0.5, 0.5], rescaled to [-1, 1] for
+ * tensors.
+ *
+ * @param stepsPerQuarter The number of quantization steps per quarter note.
+ * @param humanize If True, flatten all input velocities and microtiming. The
```
Comment (notwaldorf) [inline music/src/core/data.ts]: hmm this arg name and its description are opposite to me. Like, humanizing to me means "with microtiming", not without

Comment (notwaldorf) [reply music/src/core/data.ts]: Actually, what are these arguments that you're documenting to? If it's to the `GrooveConverterArgs`, `flattenVelocities` isn't one of them, and there's 2 other arguments that aren't documented.

File: music/src/core/data.ts [hunk 5]
```diff
@@ -799,3 +800,170 @@ export class MultitrackConverter extends DataConverter {
...
+      this.pitchClasses[c].forEach((p) => {
+        this.pitchToClass.set(p, c);
+      });
+    }
+    this.humanize = args.humanize;
+    this.splitInstruments = args.splitInstruments;
+
+    // Each drum hit is represented by 3 numbers -
+    // on/off, velocity, and offset
+    this.depth = 3;
```
Comment (notwaldorf) [inline music/src/core/data.ts]: What does depth mean though? Or, why is it called depth?

File: music/src/core/data.ts [hunk 6]
```diff
@@ -799,3 +800,170 @@ export class MultitrackConverter extends DataConverter {
...
+    const numSteps = this.numSteps || qns.totalQuantizedSteps;
+    const qpm = qns.tempos.length ? qns.tempos[0].qpm :
+                                    constants.DEFAULT_QUARTERS_PER_MINUTE;
+    const stepLength = (60. / qpm) / this.stepsPerQuarter;
+
+    const stepNotes: Array<Map<number, NoteSequence.INote>> = [];
+    for (let i = 0; i < numSteps; ++i) {
+      stepNotes.push(new Map<number, NoteSequence.INote>());
+    }
+    console.log(stepNotes[0]);
```
Comment (notwaldorf) [inline music/src/core/data.ts]: probably don't want this log in here

Comment (notwaldorf) [reply music/src/core/data.ts]: add a comment about what this block does, please 🙏

File: music/src/core/data.ts [hunk 7]
```diff
@@ -799,3 +800,170 @@ export class MultitrackConverter extends DataConverter {
...
+        const note = stepNotes[s].get(d);
+        hitVectors.set(note ? 1 : 0, s, d);
+        if (!this.humanize) {
+          velocityVectors.set(note ? note.velocity / 127 : 0, s, d);
+          offsetVectors.set(note ? getOffset(note) : 0, s, d);
+        }
+      }
+    }
+
+    // Flatten so that there is one per drum at every step, then stack the
```
Comment (notwaldorf) [inline music/src/core/data.ts]: one what per drum?

File: music/src/core/data.ts [hunk 8]
```diff
@@ -799,3 +800,170 @@ export class MultitrackConverter extends DataConverter {
...
+    const ns = NoteSequence.create({totalTime: numSteps * stepLength});
+    ns.tempos.push({qpm});
+
+    const results = await t.data() as Float32Array;
+
+    function clip(v: number, min: number, max: number) {
+      return Math.min(Math.max(v, min), max);
+    }
+    const numDrums = this.pitchClasses.length;
+    for (let s = 0; s < numSteps; ++s) {
```
Comment (notwaldorf) [inline music/src/core/data.ts]: it wouldn't suck if these 20 lines had some comments

File: music/src/core/data.ts (hunk 3 above)
Comment (adarob) [reply music/src/core/data.ts]: no, because they are needed in all subclasses

File: music/src/core/data.ts (hunk 4 above)
Comment (adarob) [reply music/src/core/data.ts]: The argument description was more relevant to the training model. Updated here to clarify.

Comment (adarob) [reply music/src/core/data.ts]: Where are you seeing flattenVelocities?

File: music/src/core/data.ts (hunk 5 above)
Comment (adarob) [reply music/src/core/data.ts]: This is variable is described in the abstract class.

**Comment by jrgillick** [comment]
Looks good! Thanks for the work on this!

**Comment by jrgillick** [comment]
👍

**Pull request merged**